### DISCUSSION TITLE
Custom Error Message For Object

### DISCUSSION THREAD

**Discussion by knoefel** [colinhacks/zod#868] [Q&A] [answered]
I'm using `zod` to validate the value of a required autocomplete field. It's working fine except that i'm not able to set a custom error message.

So when the user clears the autocomplete value it will be set to null by the autocomplete. In this case i want to print out the error message that this field is required. What i'm getting now is the generic error message:

`Expected object, received null`

Here is my schema:

```js
const projectSchema = z.object({
  id: z.string().uuid(),
  name: translationSchema,
  start: z.string(),
  end: z.string().nullable(),
});

export const ti8mProjectSchema = z.object({
  project: projectSchema,
});
```

So is there a way to define a custom error message to show when the field gets the `null` value? The only thing i found is `ZodErrorMap`, but i don't want to set it global, just for the desired fields.

Thx in advance and keep up the great work!!!

**Comment by StefanTerdell** [comment]
Hm, this is a slightly odd one, since you want the inferred type to be able to be null (so you can let the autocomplete set it to null without TS barking up the wrong tree), but let the validation fail if it is. This should do the trick:

```typescript
import { z } from 'zod';

const schema = z
  .object({ hm: z.boolean() })
  .nullable()
  .refine((x) => x !== null, 'pls no null 😿');

type Type = z.infer<typeof schema>;
//type Type = {
//    hm: boolean;
//} | null

schema.parse(null);
//ZodError: [
//    {
//      "code": "custom",
//      "message": "pls no null 😿",
//      "path": []
//    }
//  ]
```

**Reply by knoefel** [reply]
Thx a lot for your answer!

As you think it's an odd one - maybe there is something wrong in the way i'm thinking about this use case? But i think it isn't a rare case that this could happen with any kind of select/autocomplete where the user can clear the value and hit save so the validation would kick in.

In the end i went with your solution as i could add the validation right to the field validation definition instead of defining it via an outer `errorMap` configuration of the schema.

**Reply by StefanTerdell** [reply]
It's just odd in the meaning that I usually want my schemas to reflect my types exactly, but in this case I guess it's warranted :)

**Comment by JacobWeisenburger** [comment, marked as answer]
Here's 1 way to do it.
```ts
const projectSchema = z.object( {
    id: z.string().uuid(),
    name: translationSchema,
    start: z.string(),
    end: z.string().nullable(),
}, {
    errorMap: ( issue, ctx ) => {
        if ( ctx.data === null ) return { message: 'custom errorMap message' }
        return { message: ctx.defaultError }
    }
} )

const result = projectSchema.safeParse( null )
!result.success && result.error.formErrors.formErrors[ 0 ] // 'custom errorMap message'
```

___
> ...but i don't want to set it global, just for the desired fields.

What do you mean by this?

**Reply by knoefel** [reply]
Thx a lot for your answer. 

> ...but i don't want to set it global, just for the desired fields.
> 
> What do you mean by this?

I meant exactly what you suggested as a solution: scoping the `ZodErrorMap` to a specific schema. 😉
